Intermittent issues when deploying local charms

I am getting the following error when attempting to deploy local charms on a k8s model: ```ERROR Post https://jaas.ps7.canonical.com:443/k8s-jaas-ps7-jimm-jimm/model/40cad239-1fe9-497f-89a2-ce70ab3e33af/charms?revision=0&schema=local&series=: ```

The full command and debug output 
```sh
$ juju deploy --show-log --verbose ./charm/flask-app_amd64.charm webteam-juju-demos-testing-pr10 --resource flask-app-image=ghcr.io/m7mdisk/webteam-juju-demos-testing:webteam-juju-demos-testing-pr10  
07:50:40 INFO  juju.cmd supercommand.go:56 running juju [3.6.14 b08ad63ae633a35dbfda995a757dcf9db4e11160 gc go1.25.6]
07:50:40 DEBUG juju.cmd supercommand.go:57   args: []string{"/snap/juju/33784/bin/juju", "deploy", "--debug", "./flask-app_amd64.charm", "webteam-juju-demos-testing-pr10", "--resource", "flask-app-image=ghcr.io/m7mdisk/webteam-juju-demos-testing:webteam-juju-demos-testing-pr10", "-m", "795798e4-922f-49c7-9169-004ffc17df90@serviceaccount/k8s-webteam-demos-default"}
07:50:40 INFO  juju.juju api.go:90 connecting to API addresses: [jaas.ps7.canonical.com:443/k8s-jaas-ps7-jimm-jimm]
07:50:40 DEBUG juju.api apiclient.go:731 looked up jaas.ps7.canonical.com -> [91.189.92.177]
07:50:40 DEBUG juju.api apiclient.go:1089 successfully dialed "wss://jaas.ps7.canonical.com:443/k8s-jaas-ps7-jimm-jimm/model/40cad239-1fe9-497f-89a2-ce70ab3e33af/api"
07:50:40 INFO  juju.api apiclient.go:610 connection established to "wss://jaas.ps7.canonical.com:443/k8s-jaas-ps7-jimm-jimm/model/40cad239-1fe9-497f-89a2-ce70ab3e33af/api"
07:50:40 DEBUG juju.cmd.loginprovider tryinorderloginprovider.go:64 successful login using provider ClientCredentialsLoginProvider
07:50:40 INFO  juju.juju api.go:90 connecting to API addresses: [jaas.ps7.canonical.com:443/k8s-jaas-ps7-jimm-jimm]
07:50:41 DEBUG juju.api apiclient.go:731 looked up jaas.ps7.canonical.com -> [91.189.92.177]
07:50:41 DEBUG juju.api apiclient.go:1089 successfully dialed "wss://jaas.ps7.canonical.com:443/k8s-jaas-ps7-jimm-jimm/api"
07:50:41 INFO  juju.api apiclient.go:610 connection established to "wss://jaas.ps7.canonical.com:443/k8s-jaas-ps7-jimm-jimm/api"
07:50:41 DEBUG juju.cmd.loginprovider tryinorderloginprovider.go:64 successful login using provider ClientCredentialsLoginProvider
07:50:41 INFO  juju.cmd.juju.application.deployer baseselector.go:121 with the user specified base "ubuntu@24.04/stable"
07:50:41 INFO  juju.cmd.juju.application.deployer baseselector.go:162 with the latest LTS base "ubuntu@24.04/stable"
07:50:41 INFO  cmd charm.go:262 Preparing to deploy local charm: "flask-app" 
07:51:14 DEBUG juju.api monitor.go:35 RPC connection died
07:51:14 DEBUG juju.api monitor.go:35 RPC connection died
ERROR Post https://jaas.ps7.canonical.com:443/k8s-jaas-ps7-jimm-jimm/model/40cad239-1fe9-497f-89a2-ce70ab3e33af/charms?revision=0&schema=local&series=: 
07:51:14 DEBUG cmd supercommand.go:571 error stack: 
github.com/juju/juju/api.unmarshalHTTPErrorResponse:145: 
/build/snapcraft-juju-66d00790d7f3b27b764c6356cfdeb735/parts/juju/build/vendor/gopkg.in/httprequest.v1/client.go:307: Post https://jaas.ps7.canonical.com:443/k8s-jaas-ps7-jimm-jimm/model/40cad239-1fe9-497f-89a2-ce70ab3e33af/charms?revision=0&schema=local&series=
/build/snapcraft-juju-66d00790d7f3b27b764c6356cfdeb735/parts/juju/build/vendor/gopkg.in/httprequest.v1/client.go:185: 
github.com/juju/juju/api/client/charms.(*httpPutter).PutCharm:67: 
github.com/juju/juju/api/client/charms.(*fallbackPutter).PutCharm:151: 
github.com/juju/juju/api/client/charms.(*LocalCharmClient).AddLocalCharm:116: 
github.com/juju/juju/cmd/juju/application/deployer.(*localCharm).PrepareAndDeploy:272: 
```
logs from a successful run:
```
10:40:59 INFO  juju.cmd supercommand.go:56 running juju [3.6.14 b08ad63ae633a35dbfda995a757dcf9db4e11160 gc go1.25.6]
10:40:59 DEBUG juju.cmd supercommand.go:57   args: []string{"/snap/juju/33784/bin/juju", "deploy", "--debug", "./flask-app_amd64.charm", "webteam-juju-demos-testing-pr11", "--resource", "flask-app-image=ghcr.io/m7mdisk/webteam-juju-demos-testing:webteam-juju-demos-testing-pr11", "-m", "795798e4-922f-49c7-9169-004ffc17df90@serviceaccount/k8s-webteam-demos-default"}
10:40:59 INFO  juju.juju api.go:90 connecting to API addresses: [jaas.ps7.canonical.com:443/k8s-jaas-ps7-jimm-jimm]
10:40:59 DEBUG juju.api apiclient.go:731 looked up jaas.ps7.canonical.com -> [91.189.92.177]
10:40:59 DEBUG juju.api apiclient.go:1089 successfully dialed "wss://jaas.ps7.canonical.com:443/k8s-jaas-ps7-jimm-jimm/model/40cad239-1fe9-497f-89a2-ce70ab3e33af/api"
10:40:59 INFO  juju.api apiclient.go:610 connection established to "wss://jaas.ps7.canonical.com:443/k8s-jaas-ps7-jimm-jimm/model/40cad239-1fe9-497f-89a2-ce70ab3e33af/api"
10:41:00 DEBUG juju.cmd.loginprovider tryinorderloginprovider.go:64 successful login using provider ClientCredentialsLoginProvider
10:41:00 INFO  juju.juju api.go:90 connecting to API addresses: [jaas.ps7.canonical.com:443/k8s-jaas-ps7-jimm-jimm]
10:41:00 DEBUG juju.api apiclient.go:731 looked up jaas.ps7.canonical.com -> [91.189.92.177]
10:41:00 DEBUG juju.api apiclient.go:1089 successfully dialed "wss://jaas.ps7.canonical.com:443/k8s-jaas-ps7-jimm-jimm/api"
10:41:00 INFO  juju.api apiclient.go:610 connection established to "wss://jaas.ps7.canonical.com:443/k8s-jaas-ps7-jimm-jimm/api"
10:41:01 DEBUG juju.cmd.loginprovider tryinorderloginprovider.go:64 successful login using provider ClientCredentialsLoginProvider
10:41:01 INFO  juju.cmd.juju.application.deployer baseselector.go:121 with the user specified base "ubuntu@24.04/stable"
10:41:01 INFO  juju.cmd.juju.application.deployer baseselector.go:162 with the latest LTS base "ubuntu@24.04/stable"
10:41:01 INFO  cmd charm.go:262 Preparing to deploy local charm: "flask-app" 
10:41:09 INFO  cmd charm.go:282 Located local charm "flask-app", revision 23
10:41:10 INFO  cmd charm.go:136 Deploying "webteam-juju-demos-testing-pr11" from local charm "flask-app", revision 23 on ubuntu@24.04/stable
10:41:10 DEBUG juju.api monitor.go:35 RPC connection died
10:41:10 DEBUG juju.api monitor.go:35 RPC connection died
10:41:10 INFO  cmd supercommand.go:578 command finished
```

The same command often does succeed (with no change to the charm, resources, or anything else). simply retrying enough times will get it to work eventually. I would approximate it fails around 70-80% of the time.

--- a/internal/rpc/multitransport.go
+++ b/internal/rpc/multitransport.go
@@ -33,6 +33,25 @@ type multiBackendTransport struct {
 	currentURLIndex int
 }
 
+// readTrackingNoOpCloserBody wraps an io.ReadCloser to track whether it has been read from
+// and provides a no-op Close method to allow the original request body to be closed separately.
+// This is similar to the stdlib's net/http/transport.go readTrackingBody but adds a no-op closer.
+type readTrackingNoOpCloserBody struct {
+	io.ReadCloser
+	didRead bool
+}
+
+// Read tracks that the body has been read from and delegates to the underlying ReadCloser.
+func (r *readTrackingNoOpCloserBody) Read(data []byte) (int, error) {
+	r.didRead = true
+	return r.ReadCloser.Read(data)
+}
+
+// Close is a no-op to allow the original request body to be closed separately.
+func (r *readTrackingNoOpCloserBody) Close() error {
+	return nil
+}
+
 // newMultiBackendTransport creates a new MultiBackendTransport with the given
 // base transport and URLs.
 func newMultiBackendTransport(baseTransport http.RoundTripper, urls []*url.URL) (*multiBackendTransport, error) {
@@ -50,6 +69,18 @@ func newMultiBackendTransport(baseTransport http.RoundTripper, urls []*url.URL) 
 func (m *multiBackendTransport) RoundTrip(req *http.Request) (*http.Response, error) {
 	var lastErr error
 
+	// Track the original request body to determine if we can safely retry on failure
+	trackedRequest := req
+	var trackedBody *readTrackingNoOpCloserBody
+	if req.Body != nil {
+		// We do not close the original request body, this is done
+		// by the HTTP server that accepts the request, see the godoc
+		// for the http.Request Body field.
+		trackedBody = &readTrackingNoOpCloserBody{ReadCloser: req.Body}
+		trackedRequest = req.Clone(req.Context())
+		trackedRequest.Body = trackedBody
+	}
+
 	// Try each URL in succession
 	for i := 0; i < len(m.urls); i++ {
 		// Calculate the URL index to try (round-robin starting from currentURLIndex)
@@ -57,7 +88,7 @@ func (m *multiBackendTransport) RoundTrip(req *http.Request) (*http.Response, er
 		targetURL := m.urls[urlIndex]
 
 		// Clone the request to avoid modifying the original
-		clonedReq := req.Clone(req.Context())
+		clonedReq := trackedRequest.Clone(trackedRequest.Context())
 		clonedReq.URL.Scheme = targetURL.Scheme
 		clonedReq.URL.Host = targetURL.Host
 		newPath, err := url.JoinPath("/", targetURL.Path, clonedReq.URL.Path)
@@ -76,11 +107,9 @@ func (m *multiBackendTransport) RoundTrip(req *http.Request) (*http.Response, er
 			m.currentURLIndex = (urlIndex + 1) % len(m.urls)
 			return resp, nil
 		}
-		if errors.Is(err, io.EOF) {
-			// If we hit EOF, it likely means that we failed partway through
-			// the request and without the original body of the request
-			// stored, we cannot safely retry.
-			return nil, fmt.Errorf("backend failure during request: %w", err)
+		if trackedBody != nil && trackedBody.didRead {
+			// If we've read from the body we've lost some data and cannot retry, so return the error.
+			return nil, fmt.Errorf("non-retryable backend failure during request: %w", err)
 		}
 
 		// Log the failure and try the next URL

--- a/internal/rpc/multitransport_test.go
+++ b/internal/rpc/multitransport_test.go
@@ -6,6 +6,7 @@ import (
 	"io"
 	"net/http"
 	"net/http/httptest"
+	"net/http/httputil"
 	"net/url"
 	"strings"
 	"sync/atomic"
@@ -16,9 +17,9 @@ import (
 	"github.com/canonical/jimm/v3/internal/rpc"
 )
 
-// TestMultiBackendTransport tests the multiBackendTransport
-// implementation by simulating various server responses and ensuring
-// that the transport correctly handles them specifically for failover scenarios.
+// TestMultiBackendTransport exercises multiBackendTransport through a
+// reverse proxy by simulating various backend responses and checking the
+// resulting proxy behaviour for failover scenarios.
 func TestMultiBackendTransport(t *testing.T) {
 	c := qt.New(t)
 	expectedBody := "this is the request body"
@@ -56,7 +57,7 @@ func TestMultiBackendTransport(t *testing.T) {
 			http.Error(w, "unexpected number of bytes read from request body", http.StatusInternalServerError)
 			return
 		}
-		panic("foo")
+		panic("intentional panic after partial read")
 	}))
 	defer partialReadServer.Close()
 
@@ -118,22 +119,31 @@ func TestMultiBackendTransport(t *testing.T) {
 
 	for _, test := range tests {
 		c.Run(test.description, func(c *qt.C) {
-			body := strings.NewReader(expectedBody)
-			limitReader := io.LimitReader(body, int64(len(expectedBody))) // Ensure we can only read the data once.
-			req, err := http.NewRequest("POST", test.path, limitReader)
-			c.Assert(err, qt.IsNil)
-
 			transport, err := rpc.NewMultiBackendTransport(http.DefaultTransport, test.urls)
 			c.Assert(err, qt.IsNil)
 
-			resp, err := transport.RoundTrip(req)
+			proxyServer := httptest.NewServer(&httputil.ReverseProxy{
+				Rewrite:   func(pr *httputil.ProxyRequest) {},
+				Transport: transport,
+				ErrorHandler: func(w http.ResponseWriter, r *http.Request, err error) {
+					http.Error(w, err.Error(), http.StatusBadGateway)
+				},
+			})
+			defer proxyServer.Close()
+
+			body := strings.NewReader(expectedBody)
+			req, err := http.NewRequest("POST", proxyServer.URL+test.path, body)
+			c.Assert(err, qt.IsNil)
+
+			resp, err := proxyServer.Client().Do(req)
+			c.Assert(err, qt.IsNil)
+			defer resp.Body.Close()
+
 			if test.shouldError {
-				c.Assert(err, qt.IsNotNil)
+				c.Assert(resp.StatusCode, qt.Equals, http.StatusBadGateway)
 				return
 			}
 
-			c.Assert(err, qt.IsNil)
-			defer resp.Body.Close()
 			c.Assert(resp.StatusCode, qt.Equals, test.statusExpected)
 		})
 	}

